### PR TITLE
Updates for jax 0.1.41

### LINKED ISSUE #279: HMM example failing on travis with jax=0.1.41
With the lates jax, the hmm example throws an error in travis with the following message:

```
test/test_examples.py::test_cpu[hmm.py --num-samples 100 --num-warmup 100 --num-chains 2] Running:
python examples/hmm.py --num-samples 100 --num-warmup 100 --num-chains 2
Simulating data...
Starting inference...
Traceback (most recent call last):
  File "/home/travis/build/pyro-ppl/numpyro/examples/hmm.py", line 180, in <module>
    main(args)
  File "/home/travis/build/pyro-ppl/numpyro/examples/hmm.py", line 159, in main
    supervised_words, unsupervised_words,
  File "/home/travis/build/pyro-ppl/numpyro/numpyro/hmc_util.py", line 775, in initialize_model
    raise RuntimeError("Cannot find valid initial parameters. Please check your model again.")
RuntimeError: Cannot find valid initial parameters. Please check your model again.
FAILED
```

I am unable to reproduce this locally on my mac.

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR fixes some issues at the examples involving `logsumexp`. It seems that there is some issues of the upstream `logsumexp` under fastmath mode.

In addition, this incorporates various cleanup at #276.

Fix #279.

### CODE REVIEW

File: numpyro/distributions/util.py [hunk 1]
```diff
@@ -453,6 +441,12 @@ def signed_stick_breaking_tril(t):
     return y
 
 
+def logsumexp(x, axis=0, keepdims=False):
+    x_max = np.max(x, axis=axis, keepdims=True)
```
Comment (neerajprad) [inline numpyro/distributions/util.py]: Nice! Just curious - is the upstream `logsumexp` implementation doing something else that is causing issues with fastmath mode?

File: numpyro/util.py [hunk 2]
```diff
@@ -69,7 +68,7 @@ def fori_loop(lower, upper, body_fun, init_val):
             val = body_fun(i, val)
         return val
     else:
-        return jit(lax.fori_loop, static_argnums=(2,))(lower, upper, body_fun, init_val)
+        return lax.fori_loop(lower, upper, body_fun, init_val)
```
Comment (neerajprad) [inline numpyro/util.py]: Is the issue of slow compilation resolved with the latest release?

File: examples/neutra.py [hunk 3]
```diff
@@ -64,26 +67,19 @@ def main(args):
     svi_init, svi_update, _ = svi(dual_moon_model, guide, elbo, opt_init, opt_update, get_params)
     opt_state, _ = svi_init(rng_init)
 
-    def body_fn(val):
-        i, loss, opt_state_, rng_ = val
+    def body_fn(val, i):
```
Comment (neerajprad) [inline examples/neutra.py]: nit: can we keep the order (i, val) instead to match the ordering in `lax.fori_loop`'s `body_fn`?

File: numpyro/util.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/util.py]: Yes, it is fixed the last minute. :)

File: numpyro/distributions/util.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/util.py]: I think this is the only issue of NumPyro with JAX which have not been isolated yet. It is related to fastmath mode. I will ping you when I have more information. Here is the [upstream implementation](https://github.com/google/jax/blob/master/jax/scipy/special.py#L77) which does something else with -inf...

File: examples/neutra.py (hunk 3 above)
Comment (fehiepsi) [reply examples/neutra.py]: Here I want to use `lax.scan`, so i is just a dummy variable (though it is currently used in `svi_update`).

File: numpyro/distributions/util.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/distributions/util.py]: Ahh..interesting. Feel free to create a NumPyro issue with your comment, for future reference.

Comment (fehiepsi) [reply numpyro/distributions/util.py]: Actually, we have a TODO in neutra example which points to a similar issue.

**Pull request merged**